File: apps/web/src/commerce/payments.ts
// Agent payments (sandbox): turn an approved shopping plan into a spending mandate, prove consent
// with a platform passkey, and get back a signed sandbox payment intent from the API. No card data
// ever exists in the browser; the network moves no money and says so on every response.
import type { ShoppingPlan } from "./shoppingPlan";

export type PaymentLine = { productId: string; title: string; merchant: string; unitPriceUsd: string; quantity: number };
export type PaymentMandate = { maxAmountUsd: string; merchants: string[]; validForMinutes: number; budgetUsd?: string };
export type PaymentPlan = { lines: PaymentLine[]; mandate: PaymentMandate };

export type PaymentIntent = {
  intentId: string;
  status: "authorized" | "captured" | "reversed" | "declined";
  amountUsd: string;
  currency: string;
  lines: Array<PaymentLine & { totalUsd: string }>;
  merchants: string[];
  mandateMaxUsd: string;
  mandateExpiresAt: string;
  consentMethod: "passkey" | "confirm";
  credentialId: string | null;
  userVerified: boolean;
  provider: string;
  isSandbox: boolean;
  createdAt: string;
  token: string | null;
  declineCode: string | null;
  declineReason: string | null;
  history: Array<{ at: string; status: string }>;
  /** Present when a real network (Visa Acceptance test host) authorized the intent. */
  networkReference?: string | null;
  approvalCode?: string | null;
  /** Present when the external network was unreachable and the local sandbox stood in. */
  fallbackReason?: string | null;
};

/** Human label for the network that answered. */
export function providerLabel(intent: Pick<PaymentIntent, "provider">): string {
  return intent.provider === "visa-acceptance-sandbox" ? "Visa Acceptance" : "DreamGrid sandbox";
}

const usd = (n: number) => n.toFixed(2);

/** The exact plan the shopper approves: priced lines only, merchants derived from them. */
export function paymentPlanFrom(plan: ShoppingPlan, maxAmountUsd: number, validForMinutes = 24 * 60): PaymentPlan {
  const lines = plan.groups.flatMap((g) => g.lines.map((l) => ({
    productId: l.product.id, title: l.product.title, merchant: g.merchant || "Unknown store",
    unitPriceUsd: usd(l.product.priceUsd), quantity: l.quantity,
  })));
  const merchants = [...new Set(lines.map((l) => l.merchant))];
  return { lines, mandate: { maxAmountUsd: usd(maxAmountUsd), merchants, validForMinutes, ...(plan.budgetUsd > 0 ? { budgetUsd: usd(plan.budgetUsd) } : {}) } };
}

const b64 = (buf: ArrayBuffer) => btoa(String.fromCharCode(...new Uint8Array(buf))).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
const unb64 = (s: string) => Uint8Array.from(atob(s.replace(/-/g, "+").replace(/_/g, "/") + "=".repeat((4 - (s.length % 4)) % 4)), (c) => c.charCodeAt(0));

const CRED_KEY = "dreamgrid.passkey";

export type PasskeySupport = "available" | "ip-host" | "unsupported";

/** WebAuthn needs a real hostname as the relying-party ID: browsers reject IP literals like 127.0.0.1. */
export function passkeySupport(hostname = typeof location === "undefined" ? "" : location.hostname): PasskeySupport {
  if (typeof PublicKeyCredential === "undefined" || !navigator.credentials?.create) return "unsupported";
  return /^(\d{1,3}\.){3}\d{1,3}$|^\[?[0-9a-f:]+\]?$/i.test(hostname) && hostname !== "localhost" ? "ip-host" : "available";
}

/** The same page served from `localhost`, where a passkey is allowed. */
export function localhostUrl(href = location.href): string {
  const u = new URL(href);
  u.hostname = "localhost";
  return u.toString();
}

export function createPaymentsClient(fetcher: typeof fetch = fetch, storage: Pick<Storage, "getItem" | "setItem" | "removeItem"> | null = typeof localStorage === "undefined" ? null : localStorage) {
  async function request<T>(path: string, body?: unknown): Promise<T> {
    let response: Response;
    try {
      response = await fetcher(`/api/v1/payments${path}`, {
        method: body === undefined ? "GET" : "POST",
        headers: body === undefined ? undefined : { "Content-Type": "application/json" },
        body: body === undefined ? undefined : JSON.stringify(body),
        signal: AbortSignal.timeout(20_000),
      });
    } catch {
      throw new Error("The payment sandbox is unreachable. Start the API on port 8000 and try again.");
    }
    const data = await response.json().catch(() => null);
    if (!response.ok) {
      const detail = data?.detail;
      throw new Error(typeof detail === "string" ? detail : Array.isArray(detail) ? detail.map((e: { msg: string }) => e.msg).join("; ") : "The payment sandbox rejected the request.");
    }
    return data as T;
  }

  const enrolledCredential = () => storage?.getItem(CRED_KEY) || null;

  /** Create a platform passkey once per browser and enrol its public key with the API. */
  async function enrollPasskey(): Promise<string> {
    const { challenge, rpId } = await request<{ challenge: string; rpId: string }>("/passkeys/challenge", { plan: { lines: [], mandate: { maxAmountUsd: "0", merchants: ["-"], validForMinutes: 1 } }, purpose: "register" });
    const credential = await navigator.credentials.create({
      publicKey: {
        challenge: unb64(challenge),
        rp: { id: rpId, name: "DreamGrid" },
        user: { id: crypto.getRandomValues(new Uint8Array(16)), name: "shopper", displayName: "DreamGrid shopper" },
        pubKeyCredParams: [{ type: "public-key", alg: -7 }],
        authenticatorSelection: { authenticatorAttachment: "platform", residentKey: "preferred", userVerification: "required" },
        timeout: 60_000,
      },
    }) as PublicKeyCredential | null;
    if (!credential) throw new Error("No passkey was created.");
    const att = credential.response as AuthenticatorAttestationResponse;
    const authenticatorData = att.getAuthenticatorData ? att.getAuthenticatorData() : authenticatorDataFromAttestation(att.attestationObject);
    const { credentialId } = await request<{ credentialId: string }>("/passkeys", {
      challenge, clientDataJson: b64(att.clientDataJSON), authenticatorData: b64(authenticatorData),
    });
    storage?.setItem(CRED_KEY, credentialId);
    return credentialId;
  }

  /** Sign a server challenge for this exact plan; enrols first if this browser has no passkey yet. */
  async function approveWithPasskey(plan: PaymentPlan): Promise<{ challenge: string; passkey: { credentialId: string; clientDataJson: string; authenticatorData: string; signature: string } }> {
    let credentialId = enrolledCredential();
    if (!credentialId) credentialId = await enrollPasskey();
    // The sandbox keeps enrolments in memory: after an API restart the browser may hold a passkey the
    // server no longer knows. Check first and enrol a fresh one instead of failing at approval time.
    if (!(await request<{ enrolled: boolean }>(`/passkeys/${encodeURIComponent(credentialId)}`)).enrolled) {
      storage?.removeItem(CRED_KEY);
      credentialId = await enrollPasskey();
    }
    const { challenge, rpId } = await request<{ challenge: string; rpId: string }>("/passkeys/challenge", { plan, purpose: "approve" });
    const assertion = await navigator.credentials.get({
      publicKey: { challenge: unb64(challenge), rpId, allowCredentials: [{ type: "public-key", id: unb64(credentialId) }], userVerification: "required", timeout: 60_000 },
    }) as PublicKeyCredential | null;
    if (!assertion) throw new Error("The passkey prompt was dismissed.");
    const res = assertion.response as AuthenticatorAssertionResponse;
    return { challenge, passkey: { credentialId: b64(assertion.rawId), clientDataJson: b64(res.clientDataJSON), authenticatorData: b64(res.authenticatorData), signature: b64(res.signature) } };
  }

  /** Plain confirm: still a server-issued, single-use challenge bound to the plan, just no biometrics. */
  async function approveWithConfirm(plan: PaymentPlan): Promise<{ challenge: string }> {
    const { challenge } = await request<{ challenge: string }>("/passkeys/challenge", { plan, purpose: "approve" });
    return { challenge };
  }

  return {
    passkeySupport,
    hasPasskey: () => !!enrolledCredential(),
    forgetPasskey: () => storage?.removeItem(CRED_KEY),
    approveWithPasskey,
    approveWithConfirm,
    createIntent: (plan: PaymentPlan, consent: { challenge: string; passkey?: unknown }, idempotencyKey: string) =>
      request<PaymentIntent>("/intents", { plan, consent, idempotencyKey }),
    capture: (intentId: string) => request<PaymentIntent>(`/intents/${encodeURIComponent(intentId)}/capture`, {}),
    reverse: (intentId: string) => request<PaymentIntent>(`/intents/${encodeURIComponent(intentId)}/reverse`, {}),
    get: (intentId: string) => request<PaymentIntent>(`/intents/${encodeURIComponent(intentId)}`),
    ledger: () => request<PaymentIntent[]>("/intents"),
    verifyToken: (token: string) => request<{ valid: boolean; payload: Record<string, unknown> | null }>("/tokens/verify", { token }),
  };
}

export type PaymentsClient = ReturnType<typeof createPaymentsClient>;

/** Fallback for browsers without getAuthenticatorData(): read authData out of the CBOR attestation object. */
function authenticatorDataFromAttestation(attestation: ArrayBuffer): ArrayBuffer {
  const bytes = new Uint8Array(attestation);
  let pos = 0;
  const head = () => { const b = bytes[pos++]; const major = b >> 5, info = b & 0x1f; let arg = info; if (info === 24) arg = bytes[pos++]; else if (info === 25) { arg = (bytes[pos] << 8) | bytes[pos + 1]; pos += 2; } else if (info === 26) { arg = ((bytes[pos] << 24) | (bytes[pos + 1] << 16) | (bytes[pos + 2] << 8) | bytes[pos + 3]) >>> 0; pos += 4; } return { major, arg }; };
  const skip = (): void => { const { major, arg } = head(); if (major === 2 || major === 3) pos += arg; else if (major === 4) for (let i = 0; i < arg; i++) skip(); else if (major === 5) for (let i = 0; i < arg; i++) { skip(); skip(); } };
  const top = head();
  if (top.major !== 5) throw new Error("Unexpected attestation object.");
  for (let i = 0; i < top.arg; i++) {
    const key = head();
    const name = new TextDecoder().decode(bytes.subarray(pos, pos + key.arg)); pos += key.arg;
    if (name === "authData") { const v = head(); return bytes.slice(pos, pos + v.arg).buffer; }
    skip();
  }
  throw new Error("Attestation object has no authData.");
}

/** Stable idempotency key for one approval of one plan (same plan twice in a row = same intent). */
export async function planIdempotencyKey(plan: PaymentPlan, createdAt: string): Promise<string> {
  const text = JSON.stringify(plan) + createdAt;
  if (!globalThis.crypto?.subtle) return `plan-${text.length}-${createdAt}`;
  const digest = await crypto.subtle.digest("SHA-256", new TextEncoder().encode(text));
  return `plan-${[...new Uint8Array(digest)].map((b) => b.toString(16).padStart(2, "0")).join("").slice(0, 32)}`;
}

File: services/api/src/dreamgrid_api/api/v1/routes/payments.py
"""Agent payments: passkey approval, payment intents, and the sandbox ledger.

Flow: the browser asks for a challenge bound to the plan digest → the shopper signs it with a
platform passkey → the intent request carries the assertion → the network checks the mandate
and issues a signed sandbox token → capture / reverse move it through its lifecycle.
No route contacts a card network; ``provider`` and ``isSandbox`` say so on every response.
"""

from __future__ import annotations

import base64
import hashlib
import json
from datetime import UTC, datetime, timedelta
from decimal import Decimal
from typing import Annotated, Literal

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, ConfigDict, Field
from pydantic.alias_generators import to_camel

from dreamgrid_api.adapters.commerce.mock_payment_network import MockPaymentNetwork
from dreamgrid_api.adapters.commerce.passkeys import PasskeyError, PasskeyRegistry
from dreamgrid_api.boundaries.payments import (
    ConsentEvidence,
    Mandate,
    PaymentIntent,
    PaymentLine,
    PaymentNetwork,
)
from dreamgrid_api.dependencies import passkeys, payment_network, runtime_settings

router = APIRouter(prefix="/payments", tags=["payments"])


class CamelModel(BaseModel):
    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)


# ── plan + digest ─────────────────────────────────────────────────────────────


class LineIn(CamelModel):
    product_id: str = Field(min_length=1, max_length=120)
    title: str = Field(min_length=1, max_length=200)
    merchant: str = Field(min_length=1, max_length=120)
    unit_price_usd: Decimal = Field(ge=0, max_digits=10, decimal_places=2)
    quantity: int = Field(ge=1, le=99)


class MandateIn(CamelModel):
    max_amount_usd: Decimal = Field(ge=0, max_digits=10, decimal_places=2)
    merchants: list[str] = Field(min_length=1, max_length=50)
    valid_for_minutes: int = Field(default=24 * 60, ge=1, le=7 * 24 * 60)
    budget_usd: Decimal | None = Field(default=None, ge=0, max_digits=10, decimal_places=2)


class PlanIn(CamelModel):
    lines: list[LineIn] = Field(max_length=200)
    mandate: MandateIn


def plan_digest(plan: PlanIn) -> str:
    """Stable hash of exactly what the shopper is approving; the challenge is bound to it."""

    canonical = json.dumps(
        plan.model_dump(mode="json", by_alias=True), sort_keys=True, separators=(",", ":")
    )
    return hashlib.sha256(canonical.encode()).hexdigest()


# ── passkeys ─────────────────────────────────────────────────────────────────


class ChallengeRequest(CamelModel):
    plan: PlanIn
    purpose: Literal["register", "approve"] = "approve"


class ChallengeResponse(CamelModel):
    challenge: str
    plan_digest: str
    rp_id: str
    expires_at: datetime


@router.post("/passkeys/challenge", response_model=ChallengeResponse)
def passkey_challenge(
    request: ChallengeRequest,
    registry: Annotated[PasskeyRegistry, Depends(passkeys)],
) -> ChallengeResponse:
    digest = plan_digest(request.plan) if request.purpose == "approve" else "register"
    challenge = registry.issue_challenge(digest)
    return ChallengeResponse(
        challenge=challenge.value,
        plan_digest=digest,
        rp_id=registry.rp_id,
        expires_at=challenge.expires_at,
    )


class RegisterRequest(CamelModel):
    challenge: str
    client_data_json: str
    attestation_object: str | None = None
    authenticator_data: str


class RegisterResponse(CamelModel):
    credential_id: str
    created_at: datetime


def _unb64(text: str) -> bytes:
    return base64.urlsafe_b64decode(text + "=" * (-len(text) % 4))


@router.post("/passkeys", response_model=RegisterResponse, status_code=201)
def register_passkey(
    request: RegisterRequest,
    registry: Annotated[PasskeyRegistry, Depends(passkeys)],
) -> RegisterResponse:
    try:
        registry.consume_challenge(request.challenge, "register")
        key = registry.register(
            _unb64(request.client_data_json), _unb64(request.authenticator_data), request.challenge
        )
    except PasskeyError as error:
        raise HTTPException(status_code=400, detail=str(error)) from error
    return RegisterResponse(credential_id=key.credential_id, created_at=key.created_at)


class EnrolledResponse(CamelModel):
    credential_id: str
    enrolled: bool


@router.get("/passkeys/{credential_id}", response_model=EnrolledResponse)
def passkey_enrolled(
    credential_id: str, registry: Annotated[PasskeyRegistry, Depends(passkeys)]
) -> EnrolledResponse:
    """Does the in-memory sandbox still know this credential? Lets the browser re-enrol."""

    return EnrolledResponse(credential_id=credential_id, enrolled=registry.has(credential_id))


# ── intents ───────────────────────────────────────────────────────────────────


class PasskeyAssertionIn(CamelModel):
    credential_id: str
    client_data_json: str
    authenticator_data: str
    signature: str


class ConsentIn(CamelModel):
    challenge: str
    passkey: PasskeyAssertionIn | None = None
    """Absent passkey = a plain confirm click. Allowed in the sandbox, recorded as such."""


class IntentRequest(CamelModel):
    plan: PlanIn
    consent: ConsentIn
    idempotency_key: str = Field(min_length=8, max_length=120)


class LineOut(CamelModel):
    product_id: str
    title: str
    merchant: str
    unit_price_usd: str
    quantity: int
    total_usd: str


class IntentResponse(CamelModel):
    intent_id: str
    status: str
    amount_usd: str
    currency: str
    lines: list[LineOut]
    merchants: list[str]
    mandate_max_usd: str
    mandate_expires_at: datetime
    consent_method: str
    credential_id: str | None
    user_verified: bool
    provider: str
    is_sandbox: bool
    created_at: datetime
    token: str | None
    decline_code: str | None
    decline_reason: str | None
    history: list[dict[str, str]]

    @classmethod
    def from_intent(cls, intent: PaymentIntent) -> IntentResponse:
        return cls(
            intent_id=intent.intent_id,
            status=intent.status,
            amount_usd=f"{intent.amount_usd:.2f}",
            currency=intent.currency,
            lines=[
                LineOut(
                    product_id=line.product_id,
                    title=line.title,
                    merchant=line.merchant,
                    unit_price_usd=f"{line.unit_price_usd:.2f}",
                    quantity=line.quantity,
                    total_usd=f"{line.total_usd:.2f}",
                )
                for line in intent.lines
            ],
            merchants=sorted({line.merchant for line in intent.lines}),
            mandate_max_usd=f"{intent.mandate.max_amount_usd:.2f}",
            mandate_expires_at=intent.mandate.expires_at,
            consent_method=intent.consent.method,
            credential_id=intent.consent.credential_id,
            user_verified=intent.consent.user_verified,
            provider=intent.provider,
            is_sandbox=intent.is_sandbox,
            created_at=intent.created_at,
            token=intent.token,
            decline_code=intent.decline_code,
            decline_reason=intent.decline_reason,
            history=[{"at": at.isoformat(), "status": status} for at, status in intent.history],
        )


@router.post("/intents", response_model=IntentResponse, status_code=201)
def create_intent(
    request: IntentRequest,
    network: Annotated[PaymentNetwork, Depends(payment_network)],
    registry: Annotated[PasskeyRegistry, Depends(passkeys)],
) -> IntentResponse:
    digest = plan_digest(request.plan)
    try:
        registry.consume_challenge(request.consent.challenge, digest)
        if request.consent.passkey:
            pk = request.consent.passkey
            verified = registry.verify_assertion(
                pk.credential_id,
                _unb64(pk.client_data_json),
                _unb64(pk.authenticator_data),
                _unb64(pk.signature),
                request.consent.challenge,
            )
            consent = ConsentEvidence(
                method="passkey",
                credential_id=verified.credential_id,
                user_verified=verified.user_verified,
                challenge=request.consent.challenge,
            )
        else:
            consent = ConsentEvidence(
                method="confirm",
                credential_id=None,
                user_verified=False,
                challenge=request.consent.challenge,
            )
    except PasskeyError as error:
        raise HTTPException(status_code=400, detail=str(error)) from error
    lines = tuple(
        PaymentLine(
            product_id=line.product_id,
            title=line.title,
            merchant=line.merchant,
            unit_price_usd=line.unit_price_usd,
            quantity=line.quantity,
        )
        for line in request.plan.lines
    )
    mandate = Mandate(
        max_amount_usd=request.plan.mandate.max_amount_usd,
        merchants=tuple(request.plan.mandate.merchants),
        expires_at=datetime.now(UTC) + timedelta(minutes=request.plan.mandate.valid_for_minutes),
        budget_usd=request.plan.mandate.budget_usd,
    )
    intent = network.create_intent(lines, mandate, consent, idempotency_key=request.idempotency_key)
    return IntentResponse.from_intent(intent)


@router.get("/intents", response_model=list[IntentResponse])
def ledger(network: Annotated[PaymentNetwork, Depends(payment_network)]) -> list[IntentResponse]:
    return [IntentResponse.from_intent(i) for i in network.ledger()]


@router.get("/intents/{intent_id}", response_model=IntentResponse)
def get_intent(
    intent_id: str, network: Annotated[PaymentNetwork, Depends(payment_network)]
) -> IntentResponse:
    intent = network.get(intent_id)
    if intent is None:
        raise HTTPException(status_code=404, detail="Unknown payment intent.")
    return IntentResponse.from_intent(intent)


def _transition(
    network: PaymentNetwork, intent_id: str, action: Literal["capture", "reverse"]
) -> IntentResponse:
    try:
        intent = network.capture(intent_id) if action == "capture" else network.reverse(intent_id)
    except KeyError as error:
        raise HTTPException(status_code=404, detail="Unknown payment intent.") from error
    except ValueError as error:
        raise HTTPException(status_code=409, detail=str(error)) from error
    return IntentResponse.from_intent(intent)


@router.post("/intents/{intent_id}/capture", response_model=IntentResponse)
def capture(
    intent_id: str, network: Annotated[PaymentNetwork, Depends(payment_network)]
) -> IntentResponse:
    return _transition(network, intent_id, "capture")


@router.post("/intents/{intent_id}/reverse", response_model=IntentResponse)
def reverse(
    intent_id: str, network: Annotated[PaymentNetwork, Depends(payment_network)]
) -> IntentResponse:
    return _transition(network, intent_id, "reverse")


class VerifyRequest(CamelModel):
    token: str


class VerifyResponse(CamelModel):
    valid: bool
    payload: dict[str, object] | None
    provider: str
    is_sandbox: bool = True


@router.post("/tokens/verify", response_model=VerifyResponse)
def verify_token(
    request: VerifyRequest,
    network: Annotated[PaymentNetwork, Depends(payment_network)],
    settings: Annotated[object, Depends(runtime_settings)],
) -> VerifyResponse:
    """Lets a merchant-side tool (or the MCP stub) check a token the agent presents."""

    del settings
    payload = (
        network.verify_token(request.token) if isinstance(network, MockPaymentNetwork) else None
    )
    return VerifyResponse(valid=payload is not None, payload=payload, provider="dreamgrid-sandbox")
